# Saloni Patel (A4 - 4.8 - V2 Grouping Data & Aggregating Variables)

# Importing Libraries and Data

### Importing Libraries

In [3]:
# importing libraries

import pandas as pd
import numpy as np
import os

### Defining the path

In [4]:
# Creating Path for ease of code

path = r'C:\Users\psalo\OneDrive\Study\CF - DA Immersion\A4 - Python Fundamentals for Data Analyst\A4 - Instacart Basket Analysis 07-2025'


### Importing 'ords_prods_merge_post_4_7' from Prepared Data Folder

In [5]:
# Importing ords_prods_merge_post_4_7 from Prepared Data Folder

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_post_4_7.pkl'))
print(f"Number of rows imported: {ords_prods_merge.shape[0]}")

Number of rows imported: 32404859


### Checking the imported data shape

In [6]:
# Checking the ords_prods_merge_post_4_7 shape

ords_prods_merge.shape

(32404859, 19)

### Checking the imported data types

In [7]:
# Checking the ords_prods_merge Types

ords_prods_merge.dtypes

product_id                  int64
product_name               object
aisle_id                    int64
department_id               int64
prices                    float64
order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_last_order     float64
add_to_cart_order           int64
reordered                   int64
_merge                   category
price_range_loc            object
busiest_day                object
busiest_days               object
busiest_period_of_day      object
dtype: object

### Checking the imported data head

In [8]:
# Checking the ords_prods_merge Head

ords_prods_merge.head(5)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_week  \
0   3139998      138    prior            28                   6   
1   1977647      138    prior            30                   6   
2    389851      709    prior             2                   0   
3    652770      764    prior             1                   3   
4   1813452      764    prior             3                   4   

   order_hour_of_day  days_since_last_order  add_to_cart_order  reordered  \
0                 11                    3.0                  5          0   
1                 17                   20.0                  1          1   
2                 21                    6.0                 20          0   
3                 13                    7.0                 10          0   
4                 17                    9.0                 11          1   

  _merge    price_range_loc     busiest_day         busiest_days  \
0   both  Mid-range product  Regularly busy  Regularly busy days   
1   both  Mid-range product  Regularly busy  Regularly busy days   
2   both  Mid-range product     Busiest day         Busiest days   
3   both  Mid-range product  Regularly busy      Least busy days   
4   both  Mid-range product      Least busy      Least busy days   

  busiest_period_of_day  
0           Most orders  
1        Average orders  
2        Average orders  
3           Most orders  
4        Average orders

### Checking the imported data description

In [10]:
ords_prods_merge.describe()

product_id      aisle_id  department_id        prices      order_id  \
count  3.240486e+07  3.240486e+07   3.240486e+07  3.240486e+07  3.240486e+07   
mean   2.559866e+04  7.119612e+01   9.919792e+00  1.198023e+01  1.710745e+06   
std    1.408400e+04  3.821139e+01   6.281485e+00  4.956554e+02  9.872988e+05   
min    1.000000e+00  1.000000e+00   1.000000e+00  1.000000e+00  2.000000e+00   
25%    1.354400e+04  3.100000e+01   4.000000e+00  4.200000e+00  8.559470e+05   
50%    2.530200e+04  8.300000e+01   9.000000e+00  7.400000e+00  1.711049e+06   
75%    3.794700e+04  1.070000e+02   1.600000e+01  1.130000e+01  2.565499e+06   
max    4.968800e+04  1.340000e+02   2.100000e+01  9.999900e+04  3.421083e+06   

            user_id  order_number  orders_day_of_week  order_hour_of_day  \
count  3.240486e+07  3.240486e+07        3.240486e+07       3.240486e+07   
mean   1.029372e+05  1.714230e+01        2.738867e+00       1.342515e+01   
std    5.946610e+04  1.753532e+01        2.090077e+00       4.246380e+00   
min    1.000000e+00  1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.142200e+04  5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.026160e+05  1.100000e+01        3.000000e+00       1.300000e+01   
75%    1.543890e+05  2.400000e+01        5.000000e+00       1.600000e+01   
max    2.062090e+05  9.900000e+01        6.000000e+00       2.300000e+01   

       days_since_last_order  add_to_cart_order     reordered  
count           3.240486e+07       3.240486e+07  3.240486e+07  
mean            1.084114e+01       8.352547e+00  5.895873e-01  
std             8.552435e+00       7.127071e+00  4.919087e-01  
min             0.000000e+00       1.000000e+00  0.000000e+00  
25%             5.000000e+00       3.000000e+00  0.000000e+00  
50%             7.000000e+00       6.000000e+00  1.000000e+00  
75%             1.400000e+01       1.100000e+01  1.000000e+00  
max             3.000000e+01       1.450000e+02  1.000000e+00

### Replacing outliers in 'price' with Nan values as this is going to throw off further analysis

In [12]:
# Retriving all unique values of the price column in 
# sorted from small to large and we can see that there are two large numbers over 5127 rows 
# that is throwing of both charts

formatted_prices = sorted([float(f"{price:.2f}") for price in ords_prods_merge['prices'].unique()])

print(formatted_prices)

[1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2.0, 2.1, 2.2, 2.3, 2.4, 2.5, 2.6, 2.7, 2.8, 2.9, 3.0, 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9, 4.0, 4.1, 4.2, 4.3, 4.4, 4.5, 4.6, 4.7, 4.8, 4.9, 5.0, 5.1, 5.2, 5.3, 5.4, 5.5, 5.6, 5.7, 5.8, 5.9, 6.0, 6.1, 6.2, 6.3, 6.4, 6.5, 6.6, 6.7, 6.8, 6.9, 7.0, 7.1, 7.2, 7.3, 7.4, 7.5, 7.6, 7.7, 7.8, 7.9, 8.0, 8.1, 8.2, 8.3, 8.4, 8.5, 8.6, 8.7, 8.8, 8.9, 9.0, 9.1, 9.2, 9.3, 9.4, 9.5, 9.6, 9.7, 9.8, 9.9, 10.0, 10.1, 10.2, 10.3, 10.4, 10.5, 10.6, 10.7, 10.8, 10.9, 11.0, 11.1, 11.2, 11.3, 11.4, 11.5, 11.6, 11.7, 11.8, 11.9, 12.0, 12.1, 12.2, 12.3, 12.4, 12.5, 12.6, 12.7, 12.8, 12.9, 13.0, 13.1, 13.2, 13.3, 13.4, 13.5, 13.6, 13.7, 13.8, 13.9, 14.0, 14.1, 14.2, 14.3, 14.4, 14.5, 14.6, 14.7, 14.8, 14.9, 15.0, 15.1, 15.2, 15.3, 15.4, 15.5, 15.6, 15.7, 15.8, 15.9, 16.0, 16.1, 16.2, 16.3, 16.4, 16.5, 16.6, 16.7, 16.8, 16.9, 17.0, 17.1, 17.2, 17.3, 17.4, 17.5, 17.6, 17.7, 17.8, 17.9, 18.0, 18.1, 18.2, 18.3, 18.4, 18.5, 18.6, 18.7, 18.8, 18.9, 19.0, 19.

In [13]:
# for now we will replace the outliers with NaNs 

ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [14]:
# Checking if the replacement to NaN was successful

formatted_prices = sorted([float(f"{price:.2f}") for price in ords_prods_merge['prices'].unique()])

print(formatted_prices)

[1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2.0, 2.1, 2.2, 2.3, 2.4, 2.5, 2.6, 2.7, 2.8, 2.9, 3.0, 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9, 4.0, 4.1, 4.2, 4.3, 4.4, 4.5, 4.6, 4.7, 4.8, 4.9, 5.0, 5.1, 5.2, 5.3, 5.4, 5.5, 5.6, 5.7, 5.8, 5.9, 6.0, 6.1, 6.2, 6.3, 6.4, 6.5, 6.6, 6.7, 6.8, 6.9, 7.0, 7.1, 7.2, 7.3, 7.4, 7.5, 7.6, 7.7, 7.8, 7.9, 8.0, 8.1, 8.2, 8.3, 8.4, 8.5, 8.6, 8.7, 8.8, 8.9, 9.0, 9.1, 9.2, 9.3, 9.4, 9.5, 9.6, 9.7, 9.8, 9.9, 10.0, 10.1, 10.2, 10.3, 10.4, 10.5, 10.6, 10.7, 10.8, 10.9, 11.0, 11.1, 11.2, 11.3, 11.4, 11.5, 11.6, 11.7, 11.8, 11.9, 12.0, 12.1, 12.2, 12.3, 12.4, 12.5, 12.6, 12.7, 12.8, 12.9, 13.0, 13.1, 13.2, 13.3, 13.4, 13.5, 13.6, 13.7, 13.8, 13.9, 14.0, 14.1, 14.2, 14.3, 14.4, 14.5, 14.6, 14.7, 14.8, 14.9, 15.0, 15.1, 15.2, 15.3, 15.4, 15.5, 15.6, 15.7, 15.8, 15.9, 16.0, 16.1, 16.2, 16.3, 16.4, 16.5, 16.6, 16.7, 16.8, 16.9, 17.0, 17.1, 17.2, 17.3, 17.4, 17.5, 17.6, 17.7, 17.8, 17.9, 18.0, 18.1, 18.2, 18.3, 18.4, 18.5, 18.6, 18.7, 18.8, 18.9, 19.0, 19.

# A4 - 4.8 - Lesson Practice

### 6 - Creating group by with Panda

In [15]:
# group by 'product_name'

df.groupby('product_name')


NameError: name 'df' is not defined

### 7 - Aggregating Data with agg()

In [20]:
# Split the data into groups based on “department_id.”
# Apply the agg() function to each group to obtain the mean values for the “order_number” column.
                                                                
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.577493
2                17.320781
3                16.084944
4                17.530458
5                14.763075
6                16.658449
7                17.031590
8                15.076662
9                15.447580
10               18.681852
11               15.447411
12               14.327957
13               16.548642
14               16.960241
15               16.121948
16               17.803851
17               15.593633
18               19.674252
19               16.899756
20               16.255442
21               25.535479

In [21]:
#Another way to appregate 

df.groupby('department_id').order_number.mean()

department_id
1     15.577493
2     17.320781
3     16.084944
4     17.530458
5     14.763075
6     16.658449
7     17.031590
8     15.076662
9     15.447580
10    18.681852
11    15.447411
12    14.327957
13    16.548642
14    16.960241
15    16.121948
16    17.803851
17    15.593633
18    19.674252
19    16.899756
20    16.255442
21    25.535479
Name: order_number, dtype: float64

In [22]:
# Performing Multiple Aggregations

df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.577493   1  99
2                17.320781   1  96
3                16.084944   1  99
4                17.530458   1  99
5                14.763075   1  99
6                16.658449   1  99
7                17.031590   1  99
8                15.076662   1  98
9                15.447580   1  99
10               18.681852   1  99
11               15.447411   1  99
12               14.327957   1  99
13               16.548642   1  99
14               16.960241   1  99
15               16.121948   1  99
16               17.803851   1  99
17               15.593633   1  99
18               19.674252   1  99
19               16.899756   1  99
20               16.255442   1  99
21               25.535479   1  98

### 8 - Aggregating Data with transform()

### To create your flag, you’ll need some criteria. You can use the following:

If the maximum orders the user has made is over 40, then the customer will be labeled a “Loyal customer.”
If the maximum orders the user has made is over 10 but less than or equal to 40, then the customer will be labeled a “Regular customer.”
If the maximum orders the user has made is less than or equal to 10, then the customer will be labeled a “New customer.”

let’s map this task onto the three-step process introduced earlier:

1. Split the data into groups based on the “user_id” column.
2. Apply the transform() function on the “order_number” column to generate the maximum orders for each user.
3. Create a new column, “max_order,” into which you’ll place the results of your aggregation.

In [23]:
# ords_prods_merge['max_order'] = >>>> creating a new column 

# ords_prods_merge.groupby(['user_id'])['order_number'] >>>> group by user_id and order_number. so it will show which user_id has how many orders

# .transform(np.max) >>> transform the data 

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\psalo\AppData\Local\Temp\ipykernel_21008\1765849870.py:7: FutureWarning: The provided callable <function max at 0x000001CE1079AAC0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [25]:
# check the newly added column using head

ords_prods_merge.head(10)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   
5           1  Chocolate Sandwich Cookies        61             19     5.8   
6           1  Chocolate Sandwich Cookies        61             19     5.8   
7           1  Chocolate Sandwich Cookies        61             19     5.8   
8           1  Chocolate Sandwich Cookies        61             19     5.8   
9           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_week  \
0   3139998      138    prior            28                   6   
1   1977647      138    prior            30                   6   
2    389851      709    prior             2                   0   
3    652770      764    prior             1                   3   
4   1813452      764    prior             3                   4   
5   1701441      777    prior            16                   1   
6   1871483      825    prior             3                   2   
7   1290456      910    prior            12                   3   
8    369558     1052    prior            10                   1   
9    589712     1052    prior            15                   1   

   order_hour_of_day  days_since_last_order  add_to_cart_order  reordered  \
0                 11                    3.0                  5          0   
1                 17                   20.0                  1          1   
2                 21                    6.0                 20          0   
3                 13                    7.0                 10          0   
4                 17                    9.0                 11          1   
5                  7                   26.0                  7          0   
6                 14                   30.0                  2          0   
7                 10                   30.0                  1          0   
8                 20                   19.0                  1          0   
9                 12                   15.0                  2          1   

  _merge    price_range_loc     busiest_day         busiest_days  \
0   both  Mid-range product  Regularly busy  Regularly busy days   
1   both  Mid-range product  Regularly busy  Regularly busy days   
2   both  Mid-range product     Busiest day         Busiest days   
3   both  Mid-range product  Regularly busy      Least busy days   
4   both  Mid-range product      Least busy      Least busy days   
5   both  Mid-range product  Regularly busy         Busiest days   
6   both  Mid-range product  Regularly busy  Regularly busy days   
7   both  Mid-range product  Regularly busy      Least busy days   
8   both  Mid-range product  Regularly busy         Busiest days   
9   both  Mid-range product  Regularly busy         Busiest days   

  busiest_period_of_day  max_order  
0           Most orders         32  
1        Average orders         32  
2        Average orders          5  
3           Most orders          3  
4        Average orders          3  
5        Average orders         26  
6           Most orders          9  
7           Most orders         12  
8        Average orders         20  
9           Most orders         20

### creating flag in a column using loc function

In [27]:
# creating the loyalty flag using the max_order value created in the earlier command

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [28]:
# checking the frequency 

ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [31]:
# checking the specific columns with head(). This allows for a retrival of slightly larger and specific set of data for a quick check.

ords_prods_merge[['user_id', 'loyalty_flag', 'order_number', 'max_order']].head(60)

user_id      loyalty_flag  order_number  max_order
0       138  Regular customer            28         32
1       138  Regular customer            30         32
2       709      New customer             2          5
3       764      New customer             1          3
4       764      New customer             3          3
5       777  Regular customer            16         26
6       825      New customer             3          9
7       910  Regular customer            12         12
8      1052  Regular customer            10         20
9      1052  Regular customer            15         20
10     1379      New customer             1         10
11     1480    Loyal customer            36         43
12     1480    Loyal customer            38         43
13     1480    Loyal customer            39         43
14     1494  Regular customer             6         13
15     1494  Regular customer             7         13
16     1494  Regular customer             9         13
17     1540  Regular customer             4         28
18     1540  Regular customer             5         28
19     1540  Regular customer             6         28
20     1540  Regular customer             7         28
21     1540  Regular customer             8         28
22     1540  Regular customer             9         28
23     1540  Regular customer            14         28
24     1540  Regular customer            15         28
25     1540  Regular customer            18         28
26     1540  Regular customer            19         28
27     1540  Regular customer            20         28
28     1540  Regular customer            21         28
29     1540  Regular customer            22         28
30     1540  Regular customer            24         28
31     1540  Regular customer            25         28
32     1540  Regular customer            27         28
33     1540  Regular customer            28         28
34     1598      New customer             1          4
35     1598      New customer             2          4
36     1598      New customer             3          4
37     1647      New customer             1         10
38     1647      New customer             4         10
39     1647      New customer             6         10
40     1647      New customer             8         10
41     1647      New customer             9         10
42     2833      New customer             1         10
43     2850  Regular customer             7         23
44     2857  Regular customer            24         25
45     3010  Regular customer             5         13
46     3029      New customer             1          5
47     3395      New customer             4          6
48     3904      New customer             1          3
49     4006    Loyal customer             2         42
50     4006    Loyal customer             3         42
51     4006    Loyal customer             5         42
52     4006    Loyal customer             9         42
53     4006    Loyal customer            12         42
54     4122  Regular customer            11         40
55     4141      New customer             2          9
56     4343      New customer             1          4
57     4553      New customer             3          4
58     4845  Regular customer             6         14
59     6866      New customer             1          6

# A4 - 4.8 - Assignment

### 2 - In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [16]:
# Group by department id and Aggregating by order_number mean 

ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### 3 - Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

The results are slighty varied for each department as it is expected but the variation are within acceptable margin. For example, for department_id 3 the change is nominal but for department_id 21 the change is noticable - from 25.53 to 22.90. 

### 4 - Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

#### Creating new column called max_order and groupby user_id and order_number

In [17]:
# ords_prods_merge['max_order'] = >>>> creating a new column 

# ords_prods_merge.groupby(['user_id'])['order_number'] >>>> group by user_id and order_number. so it will show which user_id has how many orders

# .transform(np.max) >>> transform the data 

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\psalo\AppData\Local\Temp\ipykernel_8832\1765849870.py:7: FutureWarning: The provided callable <function max at 0x00000249AC0F7240> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [18]:
# check the newly added column using head

ords_prods_merge.head(10)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   
5           1  Chocolate Sandwich Cookies        61             19     5.8   
6           1  Chocolate Sandwich Cookies        61             19     5.8   
7           1  Chocolate Sandwich Cookies        61             19     5.8   
8           1  Chocolate Sandwich Cookies        61             19     5.8   
9           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_week  \
0   3139998      138    prior            28                   6   
1   1977647      138    prior            30                   6   
2    389851      709    prior             2                   0   
3    652770      764    prior             1                   3   
4   1813452      764    prior             3                   4   
5   1701441      777    prior            16                   1   
6   1871483      825    prior             3                   2   
7   1290456      910    prior            12                   3   
8    369558     1052    prior            10                   1   
9    589712     1052    prior            15                   1   

   order_hour_of_day  days_since_last_order  add_to_cart_order  reordered  \
0                 11                    3.0                  5          0   
1                 17                   20.0                  1          1   
2                 21                    6.0                 20          0   
3                 13                    7.0                 10          0   
4                 17                    9.0                 11          1   
5                  7                   26.0                  7          0   
6                 14                   30.0                  2          0   
7                 10                   30.0                  1          0   
8                 20                   19.0                  1          0   
9                 12                   15.0                  2          1   

  _merge    price_range_loc     busiest_day         busiest_days  \
0   both  Mid-range product  Regularly busy  Regularly busy days   
1   both  Mid-range product  Regularly busy  Regularly busy days   
2   both  Mid-range product     Busiest day         Busiest days   
3   both  Mid-range product  Regularly busy      Least busy days   
4   both  Mid-range product      Least busy      Least busy days   
5   both  Mid-range product  Regularly busy         Busiest days   
6   both  Mid-range product  Regularly busy  Regularly busy days   
7   both  Mid-range product  Regularly busy      Least busy days   
8   both  Mid-range product  Regularly busy         Busiest days   
9   both  Mid-range product  Regularly busy         Busiest days   

  busiest_period_of_day  max_order  
0           Most orders         32  
1        Average orders         32  
2        Average orders          5  
3           Most orders          3  
4        Average orders          3  
5        Average orders         26  
6           Most orders          9  
7           Most orders         12  
8        Average orders         20  
9           Most orders         20

#### Creating Loyalty flags

In [19]:
# creating the loyalty flag using the max_order value created in the earlier command 

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [20]:
# checking the specific columns with head(). This allows for a retrival of slightly larger and specific set of data for a quick check.

ords_prods_merge[['user_id', 'loyalty_flag', 'order_number', 'max_order']].head(20)

user_id      loyalty_flag  order_number  max_order
0       138  Regular customer            28         32
1       138  Regular customer            30         32
2       709      New customer             2          5
3       764      New customer             1          3
4       764      New customer             3          3
5       777  Regular customer            16         26
6       825      New customer             3          9
7       910  Regular customer            12         12
8      1052  Regular customer            10         20
9      1052  Regular customer            15         20
10     1379      New customer             1         10
11     1480    Loyal customer            36         43
12     1480    Loyal customer            38         43
13     1480    Loyal customer            39         43
14     1494  Regular customer             6         13
15     1494  Regular customer             7         13
16     1494  Regular customer             9         13
17     1540  Regular customer             4         28
18     1540  Regular customer             5         28
19     1540  Regular customer             6         28

### 5 - Analysing spending habits of each customer loyalty flag groups

The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [21]:
# Lets look at the basic stats around prices for each loyalty flag group. The stats are good indicator of trends and that can help us determine the spending habits

ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773575  1.0  25.0
New customer      7.801206  1.0  25.0
Regular customer  7.798262  1.0  25.0

##### Note >> It shows that new customers are spending more than loyal customers on individual item as the items they are putting in baskets are slightly more expensive. Also the max price of 99,999 is most likely outlier and not a real indication of the prices

### 6 - Creating a spending flag

#### The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:

- If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
- If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

First we will have to created a new column of average price (spent) for each user_id and then we will use the above given criterias to create spending flags.

#### Creating new column for avg_prod_price and aggregating values by user_id

In [22]:
# Aggregating Data with transform() Function to create a new column

ords_prods_merge['avg_prod_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\psalo\AppData\Local\Temp\ipykernel_8832\3862805806.py:3: FutureWarning: The provided callable <function mean at 0x00000249AC0F7C40> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['avg_prod_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [23]:
# Lets check out the newly created column along with user_id. I am using drop_duplicate function to only show unique values as without it I had many duplicates in my initial run

ords_prods_merge[['user_id', 'avg_prod_price', 'max_order']].drop_duplicates().head(15)


user_id  avg_prod_price  max_order
0       138        6.935811         32
2       709        7.930208          5
3       764        4.972414          3
5       777        6.935398         26
6       825        5.957576          9
7       910        6.680000         12
8      1052        7.162500         20
10     1379        5.813636         10
11     1480        5.081633         43
14     1494        6.803797         13
17     1540        6.607122         28
34     1598        7.123529          4
37     1647        4.853521         10
42     2833        7.377143         10
43     2850        8.054153         23

#### Creating 'spending flags'

In [25]:
# Creating a 'Spending' Label Flag using loc 
ords_prods_merge.loc[ords_prods_merge['avg_prod_price'] >= 10, 'spending_flag'] = 'High spender'
ords_prods_merge.loc[ords_prods_merge['avg_prod_price'] < 10, 'spending_flag'] = 'Low spender'

In [26]:
# Value Count on newly created 'spending_flag' Column

ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     32285131
High spender      119728
Name: count, dtype: int64

In [27]:
# Lets check out the newly created column along with user_id. I am using drop_duplicate function to only show unique values as without it I had many duplicates in my initial run

ords_prods_merge[['user_id', 'avg_prod_price', 'spending_flag']].drop_duplicates().head(15)

user_id  avg_prod_price spending_flag
0       138        6.935811   Low spender
2       709        7.930208   Low spender
3       764        4.972414   Low spender
5       777        6.935398   Low spender
6       825        5.957576   Low spender
7       910        6.680000   Low spender
8      1052        7.162500   Low spender
10     1379        5.813636   Low spender
11     1480        5.081633   Low spender
14     1494        6.803797   Low spender
17     1540        6.607122   Low spender
34     1598        7.123529   Low spender
37     1647        4.853521   Low spender
42     2833        7.377143   Low spender
43     2850        8.054153   Low spender

### 7 - Creating Order Frequency Flags

#### In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [28]:
# Creating a new volumn median_days_betn_orders

ords_prods_merge['median_days_betn_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)

C:\Users\psalo\AppData\Local\Temp\ipykernel_8832\421578799.py:3: FutureWarning: The provided callable <function median at 0x00000249AC236700> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['median_days_betn_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)


In [29]:
# Lets check the new column

ords_prods_merge[['user_id', 'days_since_last_order', 'median_days_betn_orders']].head(15)

user_id  days_since_last_order  median_days_betn_orders
0       138                    3.0                      8.0
1       138                   20.0                      8.0
2       709                    6.0                      7.0
3       764                    7.0                      9.0
4       764                    9.0                      9.0
5       777                   26.0                     11.0
6       825                   30.0                     19.0
7       910                   30.0                      6.0
8      1052                   19.0                      9.5
9      1052                   15.0                      9.5
10     1379                    7.0                     20.0
11     1480                    7.0                      7.0
12     1480                    7.0                      7.0
13     1480                   12.0                      7.0
14     1494                    2.0                      7.0

In [30]:
# Creating a 'Order Frequency' Label Flag using loc
ords_prods_merge.loc[ords_prods_merge['median_days_betn_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['median_days_betn_orders'] <= 20) & (ords_prods_merge['median_days_betn_orders'] > 10), 'order_frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['median_days_betn_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [31]:
# Value Counts on 'order_frequency_flag'

ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        22790523
Regular customer          6927608
Non-frequent customer     2686728
Name: count, dtype: int64

In [32]:
# Checking newly created columns

ords_prods_merge[['user_id', 'days_since_last_order', 'median_days_betn_orders', 'order_frequency_flag']].head(15)

user_id  days_since_last_order  median_days_betn_orders  \
0       138                    3.0                      8.0   
1       138                   20.0                      8.0   
2       709                    6.0                      7.0   
3       764                    7.0                      9.0   
4       764                    9.0                      9.0   
5       777                   26.0                     11.0   
6       825                   30.0                     19.0   
7       910                   30.0                      6.0   
8      1052                   19.0                      9.5   
9      1052                   15.0                      9.5   
10     1379                    7.0                     20.0   
11     1480                    7.0                      7.0   
12     1480                    7.0                      7.0   
13     1480                   12.0                      7.0   
14     1494                    2.0                      7.0   

   order_frequency_flag  
0     Frequent customer  
1     Frequent customer  
2     Frequent customer  
3     Frequent customer  
4     Frequent customer  
5      Regular customer  
6      Regular customer  
7     Frequent customer  
8     Frequent customer  
9     Frequent customer  
10     Regular customer  
11    Frequent customer  
12    Frequent customer  
13    Frequent customer  
14    Frequent customer

# Final check and Export of the updated version of df_ords_prods_merge

### Final check

In [33]:
# Checking for the newly added columns in this assignment

ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_week  ...  \
0   3139998      138    prior            28                   6  ...   
1   1977647      138    prior            30                   6  ...   
2    389851      709    prior             2                   0  ...   
3    652770      764    prior             1                   3  ...   
4   1813452      764    prior             3                   4  ...   

     price_range_loc     busiest_day         busiest_days  \
0  Mid-range product  Regularly busy  Regularly busy days   
1  Mid-range product  Regularly busy  Regularly busy days   
2  Mid-range product     Busiest day         Busiest days   
3  Mid-range product  Regularly busy      Least busy days   
4  Mid-range product      Least busy      Least busy days   

   busiest_period_of_day max_order      loyalty_flag avg_prod_price  \
0            Most orders        32  Regular customer       6.935811   
1         Average orders        32  Regular customer       6.935811   
2         Average orders         5      New customer       7.930208   
3            Most orders         3      New customer       4.972414   
4         Average orders         3      New customer       4.972414   

  spending_flag median_days_betn_orders  order_frequency_flag  
0   Low spender                     8.0     Frequent customer  
1   Low spender                     8.0     Frequent customer  
2   Low spender                     7.0     Frequent customer  
3   Low spender                     9.0     Frequent customer  
4   Low spender                     9.0     Frequent customer  

[5 rows x 25 columns]

### Exporting

In [34]:
# Exporting final df in pkl version and printing the number of rows exported for one final verification

num_rows = len(ords_prods_merge)  
print(f"{num_rows} rows exported to pkl file.")

ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_post_4_8.pkl'))

32404859 rows exported to pkl file.
